docs/RFCS/20190722_error_handling.md

```diff
@@ -0,0 +1,188 @@
+- scope: which scope does this error happen, including upstream, downstream, DM inner
+- level: emergency level for this error, including high, medium and low
+- args: used for message generator, such as we have an error ErrFailedFlushCheckpoint = terror.Syncer.New(5300, “failed to flush checkpoint %s”), we can use this error like ErrFailedFlushCheckpoint.GenWithArgs(checkpoint), so we don’t need additional error messages when we use this error
+- rawCause: used to record some root error from the third party function call
+- stack: thanks to [pingcap/errors/StackTracer](https://www.google.com/url?q=https://github.com/pingcap/errors/blob/master/stack.go%23L13-L17&sa=D&ust=1563783859452000), we can use this to record stack trace easily
+
+```go
+type ErrCode int
+type ErrClass int
+type ErrScope int
+type ErrLevel int
+
+type Error struct {
+	code		ErrCode
+	class		ErrClass
+	scope		ErrScope
+	level		ErrLevel
+	message		string
+	args		[]interface{}
+	rawCause	error
+	stack		errors.StackTracer
+}
+```
+
+#### Error classification and error codes
+
+1. Error is classfied by class field, which relates to code logic
+2. Error codes range allocation will be added later
+
+### Goal of error handling
+
+#### Provide better error equivalences check
+
+1. Enable fast, reliable and secure determination of whether a particular error cause is present（no relying on the presence of a substring in the error messages）
+2. Provide the following interface for error equivalences check.
+3. Support protobuf-encodable error object，so we can work with errors transmitted across the network via GRPC.（TODO）
+
+```go
+// Equal returns true iff the error contains `reference` in any of its
+func (e *Error) Equal(reference error) bool {}
+
+// EqualAny is like Equal() but supports multiple reference errors.
+func (e terror) EqualAny(references ...error) bool {}
+```
+
+#### Enhance the error chain
+
+1. When we generate a new error in DM level source, we always use `Generate` or `Generatef` to create a new Error instance from a defined error list.
+2. When we invoke a third party function and get an error, we should change this error to adapt our error system, we have two choices here:
+
+- Keep the error message from third party function and create a related Error instance in new error system.
+- Create a new Error instance, and save the third party error in its `rawCause` field.
+
+1. Supposing one function A invokes another function B, and function A is also invoked by other code, both functions A and function B are DM level code and have an error field in their return values, we should make a rule about how to propagate error to upper code. In this scenario, we call function A as current function, we call function B as inner function, and we call the code invokes function A as upper code stack. The inner function returns `err != nil`, the current function shall propagate this error to the upper code stack, it will generate different error object based on the code logic.
```

Comment:
```suggestion
3. Supposing one function A invokes another function B, and function A is also invoked by other code, both functions A and function B are DM level code and have an error field in their return values, we should make a rule about how to propagate error to upper code. In this scenario, we call function A as current function, we call function B as inner function, and we call the code invokes function A as upper code stack. The inner function returns `err != nil`, the current function shall propagate this error to the upper code stack, it will generate different error object based on the code logic.
```